Feat: Custom plugins reference (#1250)
Kong/developer.konghq.com: app/custom-plugins/installation-and-distribution.md

Reviewer: Should this be linked to from https://deploy-preview-1250--kongdeveloper.netlify.app/custom-plugins/

Author: Yes, I added a link

related_resources:
  - text: Custom plugins
    url: /custom-plugins/
  - text: Custom plugins reference
    url: /custom-plugins/reference/
---

Custom plugins for {{site.base_gateway}} consist of Lua source files that need to be in the file system of each of your {{site.base_gateway}} nodes. 
This guide will provide you with step-by-step instructions that will make a {{site.base_gateway}} node aware of your custom plugin(s).

These steps should be applied to each node in your {{site.base_gateway}} cluster, to ensure the custom plugins are available on each one of them.

## Package sources

To package your custom plugin code, we recommend LuaRocks, as it is installed along with {{site.base_gateway}} when using one of the official distribution packages. 
However, you can also use a regular packing strategy, such as `tar`.

When using LuaRocks, you must create a `rockspec` file, which specifies the package contents. 
For an example, see the [{{site.base_gateway}} plugin template](https://github.com/Kong/kong-plugin). 
For more information about the format, see the LuaRocks [documentation on rockspecs](https://github.com/keplerproject/luarocks/wiki/Creating-a-rock).

> Pack your rock using the following commands (from the plugin repo):

Reviewer: ```suggestion
To package your plugin as a rock, run the following commands from within the plugin repository:
```

   After this step it will be impossible to re-apply the plugin. 
   This step requires to restart or reload the {{site.base_gateway}} node.

1. To remove the plugin thoroughly, delete the plugin-related files from each of the {{site.base_gateway}} nodes. 
   Before deleting the files, make sure to have completed the previous step, including restarting or reloading {{site.base_gateway}}. 
   If you used LuaRocks to install the plugin, you can do `luarocks remove {plugin-name}` to remove it.

## Distribute your plugin

Depending on the platform that {{site.base_gateway}} is running on, there are different ways of distributing custom plugins.

### LuaRocks

You can use [LuaRocks](https://luarocks.org/), a package manager for Lua modules. 
Your module doesn't have to live inside the {{site.base_gateway}} repository, but it can if that's how you'd like to maintain your {{site.base_gateway}} setup.

By defining your modules (and their eventual dependencies) in a [rockspec](https://github.com/keplerproject/luarocks/wiki/Creating-a-rock) file, you can install those modules on your platform via LuaRocks. 
You can also upload your module on LuaRocks and make it available to anyone.

> Here is an [example rockspec](https://github.com/Kong/kong-plugin/blob/master/kong-plugin-myplugin-0.1.0-1.rockspec) using the `builtin` build type to define modules in Lua notation and their corresponding file.

Reviewer: ```suggestion
This [example rockspec](https://github.com/Kong/kong-plugin/blob/master/kong-plugin-myplugin-0.1.0-1.rockspec) uses the `builtin` build type to define modules in Lua notation and their corresponding file.
```
